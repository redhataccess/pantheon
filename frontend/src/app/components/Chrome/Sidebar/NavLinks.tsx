import React, { Component } from 'react';
import { NavItem, NavExpandable, NavList } from '@patternfly/react-core';
import { Link } from "react-router-dom";
import fetch from 'isomorphic-fetch';

// BASE is used in the fetch call to check if isLoggedIn or isAdmin. It currently breaks the Navlinks.
// only search is displayed when BASE is consumed.
// const BASE = process.env.BROWSER? '': `http://localhost`;
class NavLinks extends Component<any, any> {

  constructor(props) {
    super(props)
    this.state = {
      activeGroup: '',
      activeItem: '',
      gotUserInfo: false,
      isAdmin: false,
      isDropdownOpen: false,
      isKebabDropdownOpen: false,
      isLoggedIn: false,
    }
  }

  public componentDidMount() {
    if (!this.state.isLoggedIn) {
      fetch("/system/sling/info.sessionInfo.json")
        .then(response => response.json())
        .then(responseJSON => {
          if (responseJSON.userID !== 'anonymous') {
            this.setState({ moduleText: 'New Module' })
            this.setState({ isLoggedIn: true })
          }
          if (responseJSON.userID === 'admin') {
            this.setState({ isAdmin: true })
          }
        })
    }
    this.setState({ activeGroup: 'grp-1', activeItem: 'grp-1_itm-1' });
  }
  public render() {
    const browserText = 'Content Browser'
    const consoleText = 'Web Console'
    const gitText = 'Git Import'
    const moduleText = 'New Module'
    const productText = 'New Product'
    const productsText = 'Product Listing'
    const searchText = 'Search'
    const slingHomeText = 'Sling Welcome'

    return (
      <React.Fragment>
        <NavList>
          <NavExpandable title="Modules" groupId="grp-1" isActive={this.state.activeGroup === 'grp-1'} isExpanded={true} data-testid="navLink_modules">
            <NavItem groupId="grp-1" itemId="grp-1_itm-1" isActive={this.state.activeItem === 'grp-1_itm-1'}>
              <Link to='/search' data-testid="navLink_search">{searchText}</Link>
            </NavItem>
            {(this.state.isLoggedIn) && (<NavItem groupId="grp-1" itemId="grp-1_itm-2" isActive={this.state.activeItem === 'grp-1_itm-2'}>
              <Link to='/module'>{moduleText}</Link>
            </NavItem>)}
            {(this.state.isLoggedIn) && (<NavItem groupId="grp-1" itemId="grp-1_itm-3" isActive={this.state.activeItem === 'grp-1_itm-3'}>
              <Link to='/git'>{gitText}</Link>
            </NavItem>)}
          </NavExpandable>
          {(this.state.isLoggedIn) && (<NavExpandable title="Products" groupId="grp-2" isActive={this.state.activeGroup === 'grp-2'}>
            <NavItem groupId="grp-2" itemId="grp-2_itm-1" isActive={this.state.activeItem === 'grp-2_itm-1'}>
              <Link to='/products'>{productsText}</Link>
            </NavItem>
            {(productText.length > 0) && (<NavItem groupId="grp-2" itemId="grp-2_itm-2" isActive={this.state.activeItem === 'grp-2_itm-2'}>
              <Link to='/product'>{productText}</Link>
            </NavItem>)}
          </NavExpandable>)}
          {(this.state.isLoggedIn) && (this.state.isAdmin) && (<NavExpandable title="Admin Panel" groupId="grp-3" isActive={this.state.activeGroup === 'grp-3'}>
            <NavItem groupId="grp-3" itemId="grp-3_itm-1" isActive={this.state.activeItem === 'grp-3_itm-1'} preventDefault={false} component='a'>
              <a href='/starter/index.html' target='_blank'>{slingHomeText}</a>
            </NavItem>
            <NavItem groupId="grp-3" itemId="grp-3_itm-2" isActive={this.state.activeItem === 'grp-3_itm-2'} preventDefault={false} component='a'>
              <a href='/bin/browser.html' target='_blank'>{browserText}</a>
            </NavItem>
            <NavItem groupId="grp-3" itemId="grp-3_itm-3" isActive={this.state.activeItem === 'grp-3_itm-3'} preventDefault={false} component='a'>
              <a href='/system/console/bundles.html' target='_blank'>{consoleText}</a>
            </NavItem>
          </NavExpandable>)}
        </NavList>
      </React.Fragment>
    );
  }
}

export { NavLinks }