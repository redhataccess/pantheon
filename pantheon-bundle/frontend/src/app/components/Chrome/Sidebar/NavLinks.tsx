import React, { Component } from 'react'
import { NavItem, NavExpandable, NavList } from '@patternfly/react-core'
import { Link } from "react-router-dom"
import { IAppState } from '@app/app'

// BASE is used in the fetch call to check if isLoggedIn or isAdmin. It currently breaks the Navlinks.
// only search is displayed when BASE is consumed.
// const BASE = process.env.BROWSER? '': `http://localhost`;
class NavLinks extends Component<IAppState, any> {

  constructor(props) {
    super(props)
    this.state = {
      activeGroup: '',
      activeItem: '',
      expandableClickedGroup: '',
      expandableClickedItem: '',
      gotUserInfo: false,
      isDropdownOpen: false,
      isKebabDropdownOpen: false
    }
  }
  public render() {
    const browserText = 'Content Browser'
    const consoleText = 'Web Console'
    const gitText = 'Git Import'
    const productText = 'New Product'
    const productsText = 'Product Listing'
    const searchText = 'Search'
    const slingHomeText = 'Sling Welcome'

    return (
      <React.Fragment>
        <NavList onSelect={this.onExpandableSelect} id="nav-primary-expandable">
          <NavExpandable title="Modules" groupId="grp-1" isActive={this.state.activeGroup === 'grp-1'} isExpanded={true} data-testid="navLink_modules">
            <NavItem
              groupId="grp-1"
              itemId="grp-1_itm-1"
              isActive={this.state.activeGroup === 'grp-1' && this.state.activeItem === 'grp-1_itm-1'}
              onClick={this.handleItemOnclick}
            >
              <Link to='/search' data-testid="navLink_search">{searchText}</Link>
            </NavItem>
            {(this.props.userAuthenticated) &&
              (<NavItem
                groupId="grp-1"
                itemId="grp-1_itm-2"
                isActive={this.state.activeItem === 'grp-1_itm-2'}
                onClick={this.handleItemOnclick}
              >
                <Link to='/git'>{gitText}</Link>
              </NavItem>)}
          </NavExpandable>
          {(this.props.userAuthenticated) && (<NavExpandable title="Products" groupId="grp-2" isActive={this.state.activeGroup === 'grp-2'}>
            <NavItem
              groupId="grp-2"
              itemId="grp-2_itm-1"
              isActive={this.state.activeGroup === 'grp-2' && this.state.activeItem === 'grp-2_itm-1'}
              onClick={this.handleItemOnclick}
            >
              <Link to='/products'>{productsText}</Link>
            </NavItem>
            {(productText.length > 0) &&
              (<NavItem
                groupId="grp-2"
                itemId="grp-2_itm-2"
                isActive={this.state.activeGroup === 'grp-2' && this.state.activeItem === 'grp-2_itm-2'}
                onClick={this.handleItemOnclick}
              >
                <Link to='/product'>{productText}</Link>
              </NavItem>)}
          </NavExpandable>)}
          {(this.props.userAuthenticated) && this.props.isAdmin && (<NavExpandable title="Admin Panel" groupId="grp-3" isActive={this.state.activeGroup === 'grp-3'}>
            <NavItem
              groupId="grp-3"
              itemId="grp-3_itm-1"
              isActive={this.state.activeGroup === 'grp-3' && this.state.activeItem === 'grp-3_itm-1'}
              preventDefault={false}
              component='a'
              onClick={this.handleItemOnclick}
            >
              <a href='/starter/index.html' target='_blank'>{slingHomeText}</a>
            </NavItem>
            <NavItem
              groupId="grp-3"
              itemId="grp-3_itm-2"
              isActive={this.state.activeGroup === 'grp-3' && this.state.activeItem === 'grp-3_itm-2'}
              preventDefault={false}
              component='a'
              onClick={this.handleItemOnclick}
            >
              <a href='/bin/browser.html' target='_blank'>{browserText}</a>
            </NavItem>
            <NavItem
              groupId="grp-3"
              itemId="grp-3_itm-3"
              isActive={this.state.activeGroup === 'grp-3' && this.state.activeItem === 'grp-3_itm-3'}
              preventDefault={false}
              component='a'
              onClick={this.handleItemOnclick}
            >
              <a href='/system/console/bundles.html' target='_blank'>{consoleText}</a>
            </NavItem>
          </NavExpandable>)}
        </NavList>
      </React.Fragment>
    );
  }
  private onExpandableSelect = result => {
    this.setState({
      activeGroup: result.groupId,
      activeItem: result.itemId
    });
  };

  private handleItemOnclick = (
    e: React.FormEvent<HTMLInputElement>,
    itemId: number | string,
    groupId: number | string
  ): void => {
    this.setState({
      expandableClickedGroup: groupId,
      expandableClickedItem: itemId
    });
  };
}

export { NavLinks }