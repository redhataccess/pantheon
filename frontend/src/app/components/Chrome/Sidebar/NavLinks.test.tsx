import React from 'react';
import { NavLinks } from './NavLinks';
import { NavList, NavItem, NavExpandable } from '@patternfly/react-core';
import { HashRouter as Router } from 'react-router-dom';

import { mount, shallow } from 'enzyme';
import { Link } from 'react-router-dom';
import renderer from 'react-test-renderer';
import sinon from "sinon";
import { resolve } from 'dns';

describe('NavLinks tests', () => {
  test('should render NavLinks component', () => {
    const view = shallow(<NavLinks />);
    expect(view).toMatchSnapshot();
  });

  it('should render a NavList', () => {
    const wrapper = mount(<Router><NavLinks /></Router>);
    const navList = wrapper.find(NavList);
    expect(navList.exists()).toBe(true)
  });

  it('should render a NavItem', () => {
    const wrapper = mount(<Router><NavLinks /></Router>);
    const navItem = wrapper.find(NavItem);
    expect(navItem.exists()).toBe(true)
  });

  it('should render a Link component', () => {
    const wrapper = mount(<Router><NavLinks /></Router>);
    const navLinks = wrapper.find(Link);
    expect(navLinks.exists()).toBe(true)
  });

  it('should render an Expandable component', () => {
    const wrapper = mount(<Router><NavLinks /></Router>);
    const expandable = wrapper.find(NavExpandable);
    expect(expandable.exists()).toBe(true)
  });

  it('should contain 1 NavItem without authentication', () => {
    const wrapper = shallow(<NavLinks />);

    const items = wrapper.find(NavItem);
    expect(items).toHaveLength(1);
  });

  it('should handle state changes for isLoggedIn', () => {
    const wrapper = shallow(<NavLinks />)

    expect(wrapper.state('isLoggedIn')).toBe(false);
    wrapper.setState({ 'isLoggedIn': true })
    expect(wrapper.state('isLoggedIn')).toBe(true);

    wrapper.setState({ 'moduleText': 'New Module' });
    wrapper.setState({ 'gitText': 'Git Import' });
    const navGroup1 = wrapper.find('[groupId="grp-1"]');
    expect(navGroup1.length).toBe(4)

    const navGroup2 = wrapper.find('[groupId="grp-2"]');
    expect(navGroup2.length).toBe(3)
  });

  it('should handle state changes for isAdmin', () => {
    const wrapper = shallow(<NavLinks />)

    expect(wrapper.state('isLoggedIn')).toBe(false);
    wrapper.setState({ 'isLoggedIn': true })
    expect(wrapper.state('isLoggedIn')).toBe(true);

    expect(wrapper.state('isAdmin')).toBe(false)
    wrapper.setState({ 'isAdmin': true })
    expect(wrapper.state('isAdmin')).toBe(true)
    const navGroup3 = wrapper.find('[groupId="grp-3"]');
    expect(navGroup3.length).toBe(4)
  });

  it('test browserLink function', () => {
    const wrapper = renderer.create(<Router><NavLinks /></Router>);
    const inst = wrapper.getInstance();
    expect(inst.browserLink).toMatchSnapshot();
  });

  it('test welcomeLink function', () => {
    const wrapper = renderer.create(<Router><NavLinks /></Router>);
    const inst = wrapper.getInstance();
    expect(inst.welcomeLink).toMatchSnapshot();
  });

  it('test webConsole function', () => {
    const wrapper = renderer.create(<Router><NavLinks /></Router>);
    const inst = wrapper.getInstance();
    expect(inst.consoleLink).toMatchSnapshot();
  });

  it('test render function', () => {
    const wrapper = renderer.create(<Router><NavLinks /></Router>);
    const inst = wrapper.getInstance();
    expect(inst.render).toMatchSnapshot();
  });

  it('test checkAuth function', () => {
    const wrapper = renderer.create(<Router><NavLinks /></Router>);
    const inst = wrapper.getInstance();
    expect(inst.checkAuth).toMatchSnapshot();
  });

  it('test Admin Panel links', () => {
    jest.mock('./NavLinks', () => {
      // Require the original module to not be mocked...
      const originalModule = jest.requireActual('./NavLinks');

      return {
        __esModule: true, // Use it when dealing with esModules
        ...originalModule,
        browserLink: jest.fn().mockReturnValue('window.open("/bin/browser.html")'),
        checkAuth: jest.fn().mockReturnValue(true),
        consoleLink: jest.fn().mockReturnValue('window.open("/system/console/bundles.html")'),
        welcomeLink: jest.fn().mockReturnValue('window.open("/starter/index.html")'),
      };
    });

    const browserLink = require('./NavLinks').browserLink;
    const consoleLink = require('./NavLinks').consoleLink;
    const welcomeLink = require('./NavLinks').welcomeLink;

    expect(browserLink()).toBe('window.open("/bin/browser.html")')
    expect(consoleLink()).toBe('window.open("/system/console/bundles.html")')
    expect(welcomeLink()).toBe('window.open("/starter/index.html")')
    jest.resetAllMocks()
  });

  it('calls render function', () => {
    const render = jest.fn();
    render();
    expect(render).toHaveBeenCalled();
  });

  it('test fetch api call', async () => {
    window.fetch = jest.fn().mockImplementation(async () => {
      return new Promise((resolve, reject) => {
        resolve({
          ok: true,
          status: 200,
          json: () => new Promise((resolve, reject) => {
            resolve({
              "getUserInfo": true,
              "isAdmin": false,
              "isLoggedIn": true,
            });
          })
        });
      });
      const wrapper = await shallow(<NavLinks />)
      await wrapper.update()
      expect(wrapper.state('getUserInfo')).toBe(true)
      expect(wrapper.state('isLoggedIn')).toBe(true)
      expect(wrapper.state('isAdmin')).toBe(false)
    })
  });
});