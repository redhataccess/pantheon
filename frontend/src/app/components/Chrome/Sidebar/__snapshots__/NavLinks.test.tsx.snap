// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLinks tests should render NavLinks component 1`] = `
<Fragment>
  <NavList
    ariaLeftScroll="Scroll left"
    ariaRightScroll="Scroll right"
    className=""
    variant="default"
  >
    <NavExpandable
      className=""
      data-testid="navLink_modules"
      groupId="grp-1"
      id=""
      isActive={true}
      isExpanded={true}
      srText=""
      title="Modules"
    >
      <NavItem
        groupId="grp-1"
        isActive={true}
        itemId="grp-1_itm-1"
      >
        <Link
          data-testid="navLink_search"
          to="/search"
        >
          Search
        </Link>
      </NavItem>
      <NavItem
        groupId="grp-1"
        isActive={false}
        itemId="grp-1_itm-2"
      >
        <Link
          to="/module"
        >
          New Module
        </Link>
      </NavItem>
      <NavItem
        groupId="grp-1"
        isActive={false}
        itemId="grp-1_itm-3"
      >
        <Link
          to="/git"
        >
          Git Import
        </Link>
      </NavItem>
    </NavExpandable>
    <NavExpandable
      className=""
      groupId="grp-2"
      id=""
      isActive={false}
      isExpanded={false}
      srText=""
      title="Products"
    >
      <NavItem
        groupId="grp-2"
        isActive={false}
        itemId="grp-2_itm-1"
      >
        <Link
          to="/products"
        >
          Product Listing
        </Link>
      </NavItem>
      <NavItem
        groupId="grp-2"
        isActive={false}
        itemId="grp-2_itm-2"
      >
        <Link
          to="/product"
        >
          New Product
        </Link>
      </NavItem>
    </NavExpandable>
  </NavList>
</Fragment>
`;

exports[`NavLinks tests test browserLink function 1`] = `undefined`;

exports[`NavLinks tests test checkAuth function 1`] = `undefined`;

exports[`NavLinks tests test render function 1`] = `[Function]`;

exports[`NavLinks tests test webConsole function 1`] = `undefined`;

exports[`NavLinks tests test welcomeLink function 1`] = `undefined`;
